TokenBucket.php

```diff
@@ -0,0 +1,148 @@
+    * of seconds for when the consume will be valid will be.
+    * is returned for when the next token will be available.
+    *
+    * @param amount is an integer amount of tokens that will be attempted to be
+    * consumed
+    *
+    * @return array index 0 being a boolean whether or not a token was consumed
+    *
+    *               index 1 being a double interval of seconds when there
+    *               will be enough tokens to consume the amount requested.
+    */
+   public function consume($amount) {
+      if (!is_int($amount)) {
+         throw new InvalidArgumentException("amount must be an int");
+      }
+
+      $storedBucket = $this->getStoredBucket();
+      $tokens = $this->calculateCurrentTokens($storedBucket);
+
+      $updatedTokens = $tokens - $amount;
+      if ($updatedTokens < 0) {
+         return [false, $this->readyTime($amount, $storedBucket)];
+      }
+
+      $newBucket = new StoredBucket($updatedTokens, $this->microtime());
+      $this->storeBucket($newBucket);
+      return [true, 0];
+   }
+
+   /**
+    * @return $tokens int the total number of tokens in the bucket currently.
+    */
+   public function getTokens() {
+      $storedBucket = $this->getStoredBucket();
+      $updatedTokens = $this->calculateCurrentTokens($storedBucket);
+      return $updatedTokens;
+   }
+
+   private function getStoredBucket() {
+      $storedBucket = $this->backend->get($this->key);
+      if ($storedBucket === Backend::MISS) {
+         $storedBucket = new StoredBucket($this->rate->getTokens(),
+          $this->microtime());
+      }
+
+      return $storedBucket;
+   }
+
+   /**
+    * Updates tokens to the amount that it should be after restoring the tokens
+    * that have been regenerated from the last consume to now.
+    */
+   private function calculateCurrentTokens(StoredBucket $stored) {
+      $tokens = $stored->getTokens();
+      $lastConsume = $stored->getLastConsume();
+
+      $timeLapse = $this->microtime() - $lastConsume;
+      if ($timeLapse == 0) {
+         return $tokens;
+      }
+
+      $tokens += (int)floor($timeLapse * $this->rate->getRate());
+      // don't go over maximum tokens.
+      return min($this->rate->getTokens(), $tokens);
+   }
+
+   /**
+    * Returns the number of seconds from when the amount requested will be
+    * available.
+    *
+    * If the amount requested is higher than the maximum amouunt of
+    * tokens the bucket can hold then return null to show that it will never be
+    * ready, since it can never regenerate up to that point.
+    */
+   private function readyTime($consumeAmount, StoredBucket $stored) {
+      if (!is_int($consumeAmount)) {
+         throw new InvalidArgumentException("amount must be an int");
+      }
+
+      if ($consumeAmount > $this->rate->getTokens()) {
+         return null;
+      }
+
+      $tokens = $this->calculateCurrentTokens($stored);
+      return ($consumeAmount - $tokens) / max(1, $this->rate->getRate());
```

Comment:
Why is this `max(1` here? It makes this function not work correctly for any tokens/second calculation that is < 1